registry/registry_control.go

```diff
@@ -0,0 +1,242 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package registry
+
+import (
+	"fmt"
+	"regexp"
+	"slices"
+	"strings"
+)
+
+var (
+	validAccountID = regexp.MustCompile("^(?:[a-z0-9A-Z]{32}|[-a-z0-9]{2,28})$")
+)
+
+type AuthenticationMethod int
+
+const (
+	OperatorKey AuthenticationMethod = iota
+	Store
+)
+
+var authMethodMap = map[string]AuthenticationMethod{
+	"operator-key": OperatorKey,
+	"store":        Store,
+}
+
+// StringToAuthenticationMethod converts a string to AuthenticationMethod.
+func StringToAuthenticationMethod(s string) (AuthenticationMethod, error) {
+	if method, ok := authMethodMap[s]; ok {
+		return method, nil
+	}
+	return -1, fmt.Errorf("unknown authentication method: %s", s)
+}
+
+var authMethodStringMap = map[AuthenticationMethod]string{
+	OperatorKey: "operator-key",
+	Store:       "store",
+}
+
+// String prints the string representation of AuthenticationMethod.
+func (auth AuthenticationMethod) String() string {
+	if s, ok := authMethodStringMap[auth]; ok {
+		return s
+	}
+	return "unknown"
+}
+
+// Operator holds the delegations for a single operator.
+type Operator struct {
+	OperatorID string
+	Groups     []*Group
+}
+
+// Group holds a set of views delegated through the given authentication.
+type Group struct {
+	Authentication []AuthenticationMethod
+	Views          []string
+}
+
+// groupWithView returns the group that holds the given view.
+func (operator *Operator) groupWithView(view string) (*Group, int) {
+	for _, group := range operator.Groups {
+		index, ok := slices.BinarySearch(group.Views, view)
+		if ok {
+			return group, index
+		}
+	}
+
+	return nil, 0
+}
+
+// groupWithAuthentication returns the group with the given authentication.
+// The authentication should be sorted.
+func (operator *Operator) groupWithAuthentication(authentication []AuthenticationMethod) *Group {
+	for _, group := range operator.Groups {
+		if slices.Equal(group.Authentication, authentication) {
+			return group
+		}
+	}
+
+	return nil
+}
+
+// IsDelegated checks if <accountID>/<registry>/<view> is delegated to
+// the operator under the authentication method <auth>.
+func (operator *Operator) IsDelegated(view string, auth AuthenticationMethod) bool {
+	group, _ := operator.groupWithView(view)
+	if group == nil {
+		return false
+	}
+
+	return slices.Contains(group.Authentication, auth)
+}
+
+// Delegate delegates the given views under the provided authentication methods.
+func (operator *Operator) Delegate(views []string, authentication []AuthenticationMethod) error {
```

Comment:
I thought the API wasn't agreed on yet but maybe I missed it? Just want to make sure we're not writing and reviewing code that might have to change depending on design discussions